MCO-1808: Handle unpaid Azure marketplace images

djoshy opened the pull request:
This is just up for review purposes as it depends on a few PRs which are still a WIP. **Do not directly test this PR.**


**- What I did**
This PR adds support for Azure boot image updates. It transitions older cluster from image galleries and managed images to unpaid marketplace images. This was made possible by https://github.com/openshift/installer/pull/9329 and https://github.com/coreos/stream-metadata-go/pull/97, which enhances the coreos stream to include Azure marketplace images and updates the installer workflow to use them instead of uploading the images as an Azure blob during cluster installation.

This PR can also update existing clusters with unpaid marketplace images, which may include:
- ARO clusters, which have always used unpaid marketplace images
- Older clusters that were transitioned by this mechanism in the future, and 
- Clusters created via installer after https://github.com/openshift/installer/pull/9329 lands

 
**- How to verify it**

1. Create a cluster on an older OCP release. (4.19.z at the time of writing)
2. Upgrade to a release that contains an MCO image built off this PR.
3. Opt-in the cluster to the [`TechPreview` featureset](https://docs.redhat.com/en/documentation/openshift_container_platform/4.19/html-single/nodes/index#nodes-cluster-enabling-features-about_nodes-cluster-enabling).
4. Opt-in the for [boot image updates](https://docs.redhat.com/en/documentation/openshift_container_platform/4.16/html/machine_configuration/mco-update-boot-images) using the `MachineConfiguration` object.
5. Examine the `MachineConfiguration` object's status to see if all machinesets were reconciled successfully.
6. If successful, examine a MachineSet's `providerSpec`; you should see it with a 4.19 marketplace image. Try scaling up a node, and examining its [aleph version](https://github.com/coreos/fedora-coreos-tracker/blob/main/internals/README-internals.md#aleph-version); it should have come up with the new boot image before pivoting to the OCP release image. 
7. If not successful, let me know! :smile:

djoshy commented:
/test ?

yuqi-zhang commented on manifests/machineconfigcontroller/update-bootimages-validatingadmissionpolicy.yaml:
```diff
@@ -18,5 +18,5 @@ spec:
   validations:
-    - expression: "!has(object.spec.managedBootImages) || (has(object.spec.managedBootImages) && params.status.platformStatus.type in ['GCP','AWS','VSphere'])"
-      message: "This feature is only supported on these platforms: GCP, AWS, VSphere"
+    - expression: "!has(object.spec.managedBootImages) || (has(object.spec.managedBootImages) && params.status.platformStatus.type in ['GCP','AWS','VSphere','Azure'])"
```
The VAP applies regardless of FG right?

yuqi-zhang commented on pkg/controller/machine-set-boot-image/platform_helpers.go:
```diff
@@ -216,3 +220,89 @@ func reconcileVSphereProviderSpec(streamData *stream.Stream, arch string, infra
+// Returns whether a patch is required, the updated provider spec, and any error
+func reconcileAzureProviderSpec(streamData *stream.Stream, arch string, _ *osconfigv1.Infrastructure, providerSpec *machinev1beta1.AzureMachineProviderSpec, machineSetName string, secretClient clientset.Interface) (bool, *machinev1beta1.AzureMachineProviderSpec, error) {
+
+	if providerSpec.Image.Type == machinev1beta1.AzureImageTypeMarketplaceWithPlan {
```
this would only be OKE/OPP/OCP marketplace variants right

djoshy replied:
The stream that was added does have support for paid marketplace images, so I went ahead and added support for those as well :smile:

yuqi-zhang reviewed (commented):
Haven't tested, but the workflow makes sense to me. A couple of basic questions

djoshy commented:
Rebased and added unit tests to verify updates across the several unpaid Azure variants & migration. I haven't added units for the paid marketplace images as those are yet to be implemented in the controller, we're planning to do those as part of https://issues.redhat.com/browse/MCO-1790.

pablintino commented on pkg/controller/machine-set-boot-image/platform_helpers.go:
```diff
@@ -216,3 +245,190 @@ func reconcileVSphereProviderSpec(streamData *stream.Stream, arch string, infra
+	// will also begin to use the unpaid marketplace images(aka ARO images)
+	usesLegacyImageUpload := (currentImage.ResourceID != "")
+
+	// There are two types to consider: hyperGenV1 & hyperGenV2. This determination
+	// can be done from the existing image information, and then used to pick from the stream data.
+	//
+	// Uploaded images(legacy) have a "gen2" in the resourceID field to indicate hyperGenV2
+	//
+	// Unpaid marketplace images:
+	// - have a "v2" in the SKU field to indicate hyperGenV2
+	// - aarch64 machinesets can only use hyperGenV2 images
+	//
+	// Paid marketplace images(MCO-1790):
+	// - have a "-gen1" in the SKU field to indicate hyperGenV1
+	// - we do not support paid marketplace images for aarch64
+	var usesHyperVGen2 bool
+	switch {
+	case usesLegacyImageUpload:
+		usesHyperVGen2 = strings.Contains(currentImage.ResourceID, "gen2")
+	case providerSpec.Image.Type == machinev1beta1.AzureImageTypeMarketplaceNoPlan:
+		usesHyperVGen2 = strings.Contains(currentImage.SKU, "v2") || arch == "aarch64"
+	default:
+		usesHyperVGen2 = !strings.Contains(currentImage.SKU, "gen1")
+	}
```
nit. Having this error-less piece of code here makes readers think it's going to be inmediatly used by the next few lines and it's not. Can we move it to where `usesHyperVGen2` is consumed at line 312?

djoshy replied:
Thanks for the review, I've moved this block where you suggested :smiley:

pablintino reviewed (approved):
Just a minor nit. It looks great. Thanks for the in-depth comments you have added.
/lgtm
/hold

djoshy commented:
/test all

djoshy commented:
/test all

djoshy commented:
/unhold

/test unit
